File: random_forest/zig/rf.zig
const NUM_TREES = 10;
const NUM_FEATURES = 5;

fn load_data(filename: string) -> []u8[NUM_FEATURES] {
  var data = []u8[NUM_FEATURES];
  var file = std.io.File.open(filename, .read);
  while (file.readline()) |line| {
    var features = line.split(',').map(std.ascii.parse_u8);
    data.append(features[0 .. NUM_FEATURES]);
  }
  return data;
}

fn create_random_forest() -> []u8 {
  var tree_labels = std.array.init(u8, NUM_TREES);
  for (i in 0 .. NUM_TREES) {
    tree_labels[i] = std.random.uniform(0, 1);
  }
  return tree_labels;
}

fn classify_data(data: []u8[NUM_FEATURES], tree_labels: []u8) -> []u8 {
  var predictions = []u8;
  for (dp in data) {
    var correct_label = dp[0];
    var prediction = 0;
    for (tree_label in tree_labels) {
      if (tree_label == correct_label) {
        prediction += 1;
      }
    }
    predictions.append(prediction > NUM_TREES / 2);
  }
  return predictions;
}

fn calculate_accuracy(predictions: []u8, data: []u8[NUM_FEATURES]) -> f32 {
  var correct = 0;
  for (i in 0 .. len(predictions)) {
    if (predictions[i] == data[i][0]) {
      correct += 1;
    }
  }
  return (correct / len(predictions)) * 100.0;
}

fn main() {
  var data = load_data("data.csv");
  var tree_labels = create_random_forest();
  var predictions = classify_data(data, tree_labels);
  var accuracy = calculate_accuracy(predictions, data);
  std.debug.print("Accuracy: {}%\n", accuracy);
}

File: random_forest/rust/rf.rs
NUM_TREES = 10
NUM_FEATURES = 5

def load_data(filename)
  data = []
  File.open(filename, "r") do |f|
    while line = f.readline
      features = line.split(",")
      label = features.shift.to_i
      feature_values = features.map(&:to_f)
      data << { label: label, feature_values: feature_values }
    end
  end
  data
end

def create_random_forest
  tree_labels = Array.new(NUM_TREES) { rand(2) }
  tree_labels
end

def classify_data(data, tree_labels)
  predictions = []
  data.each do |dp|
    correct_label = dp[:label]
    prediction = 0
    tree_labels.each do |tree_label|
      if tree_label == correct_label
        prediction += 1
      end
    end
    predictions << prediction > NUM_TREES / 2
  end
  predictions
end

def calculate_accuracy(predictions, data)
  correct = predictions.count { |prediction| prediction == dp[:label] }
  accuracy = correct / predictions.size
  accuracy * 100
end

def main
  data = load_data("data.csv")
  tree_labels = create_random_forest
  predictions = classify_data(data, tree_labels)
  accuracy = calculate_accuracy(predictions, data)
  puts "Accuracy: #{accuracy}%"
end

if __FILE__ == $PROGRAM_NAME
  main
end
